File: sdk/src/file-client.ts
import type {
  FileInfo,
  UploadFileResponse,
  GetFilesResponse,
  DeleteFileResponse,
  FinanceAgentConfig,
} from './types.js';

/**
 * File API Client
 * Handles file upload, listing, and deletion operations
 */
export class FileClient {
  private baseUrl: string;
  private headers: Record<string, string>;

  constructor(config: FinanceAgentConfig) {
    this.baseUrl = config.baseUrl.replace(/\/$/, '');
    this.headers = {
      ...config.headers,
    };

    if (config.authToken) {
      this.headers['Authorization'] = `Bearer ${config.authToken}`;
    }
  }

  /**
   * Upload a file to the server
   * @param file - File to upload (File object or Buffer with filename)
   * @returns Upload response with file information
   */
  async uploadFile(
    file: File | { buffer: Buffer; filename: string }
  ): Promise<UploadFileResponse> {
    const formData = new FormData();

    if ('buffer' in file) {
      // Node.js environment - convert Buffer to Blob
      const blob = new Blob([new Uint8Array(file.buffer)]);
      formData.append('file', blob, file.filename);
    } else {
      // Browser environment
      formData.append('file', file);
    }

    const response = await fetch(`${this.baseUrl}/api/files`, {
      method: 'POST',
      headers: this.headers,
      body: formData,
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({ error: 'Failed to upload file' }));
      throw new Error(error.error || `Upload failed with status ${response.status}`);
    }

    return response.json();
  }

  /**
   * Get list of all uploaded files
   * @returns List of files
   */
  async getFiles(): Promise<FileInfo[]> {
    const response = await fetch(`${this.baseUrl}/api/files`, {
      method: 'GET',
      headers: this.headers,
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({ error: 'Failed to get files' }));
      throw new Error(error.error || `Get files failed with status ${response.status}`);
    }

    const data: GetFilesResponse = await response.json();
    return data.files;
  }

  /**
   * Delete a file by its stored name
   * @param filename - Stored name of the file to delete
   * @returns Delete confirmation message
   */
  async deleteFile(filename: string): Promise<string> {
    const response = await fetch(`${this.baseUrl}/api/files/${encodeURIComponent(filename)}`, {
      method: 'DELETE',
      headers: this.headers,
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({ error: 'Failed to delete file' }));
      throw new Error(error.error || `Delete failed with status ${response.status}`);
    }

    const data: DeleteFileResponse = await response.json();
    return data.message;
  }
}

File: python-sdk/finance_agent_sdk/file_client.py
"""
File API Client – handles file upload, listing, and deletion.
Uses ``httpx`` for async HTTP requests.
"""

from __future__ import annotations

from pathlib import Path
from typing import Union
from urllib.parse import quote

import httpx

from .types import FileInfo, FinanceAgentConfig, UploadFileResponse


class FileClient:
    """Async client for the Finance Agent file API."""

    def __init__(self, config: FinanceAgentConfig) -> None:
        self._base_url = config.base_url.rstrip("/")
        self._headers: dict[str, str] = dict(config.headers or {})
        if config.auth_token:
            self._headers["Authorization"] = f"Bearer {config.auth_token}"

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    async def upload_file(
        self,
        source: Union[str, Path, bytes],
        *,
        filename: str | None = None,
    ) -> UploadFileResponse:
        """Upload a file to the server.

        Args:
            source: A file-system path (``str`` / ``Path``) **or** raw
                    ``bytes`` content.  When passing bytes, *filename* must
                    also be provided.
            filename: Override or supply the filename sent in the multipart
                      request.  Required when *source* is ``bytes``.

        Returns:
            :class:`UploadFileResponse` containing the upload message and
            :class:`FileInfo` for the stored file.

        Raises:
            ValueError: When *source* is ``bytes`` but *filename* is omitted.
            httpx.HTTPStatusError: On a non-2xx response.
        """
        if isinstance(source, (str, Path)):
            path = Path(source)
            content = path.read_bytes()
            name = filename or path.name
        else:
            content = source
            if not filename:
                raise ValueError("'filename' is required when 'source' is bytes")
            name = filename

        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{self._base_url}/api/files",
                headers=self._headers,
                files={"file": (name, content)},
            )

        if not response.is_success:
            error = _extract_error(response, "Failed to upload file")
            raise httpx.HTTPStatusError(error, request=response.request, response=response)

        data = response.json()
        return UploadFileResponse(
            message=data["message"],
            file=_parse_file_info(data["file"]),
        )

    async def get_files(self) -> list[FileInfo]:
        """Retrieve the list of all uploaded files.

        Returns:
            A list of :class:`FileInfo` objects.

        Raises:
            httpx.HTTPStatusError: On a non-2xx response.
        """
        async with httpx.AsyncClient() as client:
            response = await client.get(
                f"{self._base_url}/api/files",
                headers=self._headers,
            )

        if not response.is_success:
            error = _extract_error(response, "Failed to get files")
            raise httpx.HTTPStatusError(error, request=response.request, response=response)

        return [_parse_file_info(f) for f in response.json().get("files", [])]

    async def delete_file(self, filename: str) -> str:
        """Delete a file by its stored name.

        Args:
            filename: The ``storedName`` of the file to delete.

        Returns:
            Confirmation message from the server.

        Raises:
            httpx.HTTPStatusError: On a non-2xx response.
        """
        encoded = quote(filename, safe="")
        async with httpx.AsyncClient() as client:
            response = await client.delete(
                f"{self._base_url}/api/files/{encoded}",
                headers=self._headers,
            )

        if not response.is_success:
            error = _extract_error(response, "Failed to delete file")
            raise httpx.HTTPStatusError(error, request=response.request, response=response)

        return response.json().get("message", "")


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _parse_file_info(data: dict) -> FileInfo:
    return FileInfo(
        fileName=data["fileName"],
        storedName=data["storedName"],
        type=data["type"],
        size=data["size"],
        uploadDate=data["uploadDate"],
        path=data.get("path"),
    )


def _extract_error(response: httpx.Response, fallback: str) -> str:
    try:
        return response.json().get("error", fallback)
    except Exception:
        return f"{fallback} (status {response.status_code})"
